### ISSUE TITLE
DefectDojo findings are not synced back correctly

### ISSUE THREAD

**Issue by moxli** [secureCodeBox/secureCodeBox#2057]
### Describe the bug

I am seeing issues with my synced back findings.

My setup includes DefectDojo, the DefectDojo persistence provider, the notification hook and the ZAP + Nikto scanners.

I am trying to prevent a notification from being sent in case all findings in are already "handled" (mitigated as false positive or risk accepted) in DefectDojo as described shortly here: https://www.securecodebox.io/docs/hooks/notification-webhook/#configuration-of-a-notification

I have three findings in DefectDojo which were found by ZAP (Automation Framework) and imported using the persistence hook.

After doing a risk acceptance I expect all of them to be marked as duplicates during the next scan and all should include the "defectdojo.org/original-finding" data.

However as you can see in the json attached the last finding is not marked as a duplicate and the original finding is not attached.

[findings-zap.json](https://github.com/secureCodeBox/secureCodeBox/files/13053912/findings-zap.json)

With Nikto I am seeing the same issues, but with two (out of five) findings not have the details included. From the picture below the Alt-SVC finding was correctly duplicated and the X-Content finding was not.

Interestingly they are also the last ones in the json! Is there any sorting going on or is it just a coincidence?

### Steps To Reproduce
1. Create DefectDojo Persistence Setup as described below
2. Run scans and import the into DefectDojo through the provider
3. Check the synced back to json for the missing details.

### Expected behavior
I expect the synced back findings to be marked as a duplicate and to include the original DefectDojo finding as described here

### System (please complete the following information):
 - secureCodeBox Version/Release: 4.1.0
 - Kubernetes Version: v1.24.14-gke.2700

### Screenshots / Logs

**ZAP Findings in DefectDojo**
![image](https://github.com/secureCodeBox/secureCodeBox/assets/26149616/65a7cbd0-0ff2-42c8-a9b9-bb771d95f24b)

**Nikto Findings in DefectDojo**
![image](https://github.com/secureCodeBox/secureCodeBox/assets/26149616/26fbec2b-ed82-45c8-8df9-7c954cf8937a)

**Persistence Hook Log**
```
 @LT100144 ~ k logs securecodebox-defectdojo-persistence-defectdojo-nikto-sch-rws67
2023-10-20 11:34:21 INFO  DefectDojoPersistenceProvider:25 - Starting DefectDojo persistence provider
2023-10-20 11:34:21 INFO  DefectDojoPersistenceProvider:36 - Downloading Scan Result
2023-10-20 11:34:22 INFO  DefectDojoPersistenceProvider:40 - Uploading Findings to DefectDojo at: https://REDACTED
2023-10-20 11:34:22 INFO  VersionedEngagementsStrategy:85 - Running with DefectDojo User Id: 12
2023-10-20 11:34:22 INFO  VersionedEngagementsStrategy:203 - Looking for ID of ProductType 'Customer Facing Infrastructure'
2023-10-20 11:34:22 INFO  VersionedEngagementsStrategy:210 - Using ProductType Id: 7
2023-10-20 11:34:24 INFO  VersionedEngagementsStrategy:115 - Uploaded Scan Report as testID 2668 to DefectDojo
2023-10-20 11:34:24 INFO  DefectDojoPersistenceProvider:44 - Identified total Number of findings in DefectDojo: 5
2023-10-20 11:34:24 INFO  DefectDojoPersistenceProvider:51 - Overwriting secureCodeBox findings with the findings from DefectDojo.
2023-10-20 11:34:25 INFO  S3Service:25 - Uploading Findings to S3
2023-10-20 11:34:26 INFO  KubernetesService:88 - Updating Scan metadata
2023-10-20 11:34:26 INFO  DefectDojoPersistenceProvider:63 - DefectDojo Persistence Completed
```

### Additional context

**Helm Setup**
```
resource "helm_release" "securecodebox-defectdojo" {
  name       = "securecodebox-defectdojo"
  repository = "https://charts.securecodebox.io"
  chart      = "persistence-defectdojo"
  namespace  = kubernetes_namespace_v1.securecodebox.metadata[0].name
  version    = "4.1.0"

  values = [
    <<-EOF
    hook:
      tolerations:
      - key: "instance_type"
        value: "preemptible"
        effect: "NoSchedule"
    EOF
  ]

  set {
    name = "defectdojo.url"
    value = "https://REDACTED"
  }

  set {
    name = "defectdojo.authentication.userSecret"
    value = kubernetes_secret_v1.defectdojo-scanner-user-credentials.metadata[0].name
  }

  set {
    name = "hook.resources.limits.memory"
    value = "500Mi"
  }
}

resource "helm_release" "securecodebox-nikto" {
  name       = "securecodebox-nikto"
  repository = "https://charts.securecodebox.io"
  chart      = "nikto"
  namespace  = kubernetes_namespace_v1.securecodebox.metadata[0].name
  version    = "4.0.1"

  values = [
    <<-EOF
    parser:
      tolerations:
      - key: "instance_type"
        value: "preemptible"
        effect: "NoSchedule"
    scanner:
      tolerations:
      - key: "instance_type"
        value: "preemptible"
        effect: "NoSchedule"
    EOF
  ]
}

resource "helm_release" "securecodebox-zap" {
  name       = "securecodebox-zap"
  repository = "https://charts.securecodebox.io"
  chart      = "zap"
  namespace  = kubernetes_namespace_v1.securecodebox.metadata[0].name
  version    = "4.0.1"

  values = [
    <<-EOF
    parser:
      tolerations:
      - key: "instance_type"
        value: "preemptible"
        effect: "NoSchedule"
    scanner:
      tolerations:
      - key: "instance_type"
        value: "preemptible"
        effect: "NoSchedule"
    EOF
  ]

  set {
    name = "scanner.resources.requests.cpu"
    value = "2200m"
  }

  set {
    name = "scanner.resources.requests.memory"
    value = "8200Mi"
  }

  set {
    name = "scanner.resources.limits.memory"
    value = "8200Mi"
  }
}

}

resource "helm_release" "securecodebox-notification" {
  name       = "securecodebox-notification"
  repository = "https://charts.securecodebox.io"
  chart      = "notification"
  namespace  = kubernetes_namespace_v1.securecodebox.metadata[0].name
  version    = "4.1.0"

  values = [
    <<-EOF
    customTemplateMap:
      exists: true
      name: "${kubernetes_config_map_v1.email-template.metadata[0].name}"
    hook:
      tolerations:
      - key: "instance_type"
        value: "preemptible"
        effect: "NoSchedule"
    notificationChannels:
      - name: email
        type: email
        template: REDACTED
        skipNotificationOnZeroFinding: true
        rules:
        - matches:
            anyOf:
            - category: "DefectDojo Imported Finding"
              attributes:
                defectdojo.org/original-finding:
                  riskAccepted: false
            - category: "DefectDojo Imported Finding"
              attributes:
                duplicate: false
        endPoint: "REDACTED"
    env:
      - name: EMAIL_FROM
        value: REDACTED
      - name: SMTP_CONFIG
        valueFrom:
          secretKeyRef:
            name: email-config
            key: smtp-config
    EOF
  ]
}


```

**Comment by J12934** [comment]
Hi @moxli 

if you go to the last zap advanced finding from the report in in defectdojo (https://REDACTED/finding/16788) 
is that really properly marked as a duplicate?

if so, not sure what might be causing this.
Maybe something where the duplicate detection wasn't completly done by defect dojo by the time the findings were refetched? 
I think there is a setting in defect dojo if the deduplication should be done asyncronously, is that maybe turned on? I don't think this is supported by the hook. (though we probably should, or at least document that we don't support it) Don't have admin access to a DefectDojo instance right now, so can't easily check.

**Comment by moxli** [comment]
@J12934  thanks your quick feedback!

Yes the finding 16788 was correctly marked as a duplicate by DefectDojo.

This is it:

![image](https://github.com/secureCodeBox/secureCodeBox/assets/26149616/37ea7186-4a1c-4a24-9237-083a326a9c58)

Regarding your question about the refetch, I also thought about this since both of my imports show the same behavior where the later results are not marked as duplicates.

Do you know when the results are refetched from DefectDojo and if this could be increased? I would not care if the pod waits for a a minute or two.

In the settings I could not find a setting regarding asyncronous deduplication but rather it seems like this is done all the time?

[Here](https://defectdojo.github.io/django-DefectDojo/getting_started/running-in-production/#celery-worker) it says "...leveraging async functions (like deduplication),.." for the celery worker. 

The setting ASYNC_FINDING_IMPORT is not turned on in my instance, but I am not sure if this affects deduplication.

**Comment by J12934** [comment]
As far as I remember they are basically immediately refetched after the import. There is no waiting time and currently no way to configure one.

Not sure why this hasn't come up before. I don't think we changed that setting either maybe the deduplication in DD got slower or just instance is slower than usual 😅

Instead of just adding a timeout it would be good to investigate if this is the acutual cause and if there is a good way to reliably wait until the deduplication for a import has finished

**Comment by moxli** [comment]
@J12934 

During my current testing there are zero other imports running on this DefectDojo. 
I am not expecting it to be slower than other deployments. It runs on beefy GKE nodes and on separate database instance with two celery workers.

I just tested the following: 

![image](https://github.com/secureCodeBox/secureCodeBox/assets/26149616/5041ffbc-d242-44ba-a248-20a53a6cac7a)

I set the max number of duplicates to zero.

A task is running every minutes and it is deleting every duplicate.

Now looking into the synced back finding json file (it includes three findings, one is marked as a duplicate) every link to a finding, even the ones which were marked as non duplicates, are no longer working because the findings were automatically deleted by the clean up task.

**Comment by moxli** [comment]
@J12934 as another indicator, every time I run a ZAP scan which results in these three findings it's basically a 50% chance for 2/3 and 50% chance for 1/3 to be marked as a duplicate.

**Comment by moxli** [comment]
@J12934 

As far as I can tell this is the celery task responsible for handling the imported finding (I am not 100% sure since there are four unique ids, even though there are only three findings).
The last post_processing task finishes at 15:02:03, which is "at the same time" as the persistence pod logs that it will overwrite the json now...

```
defectdojo-celery-worker-675b4c8b8b-229bd celery [20/Oct/2023 15:02:03] INFO [celery.worker.strategy:161] Task dojo.finding.helper.post_process_finding_save[713d4294-4553-4f04-bf9d-486f1b81eb0a] received
defectdojo-celery-worker-675b4c8b8b-229bd celery [20/Oct/2023 15:02:03] INFO [celery.app.trace:131] Task dojo.finding.helper.post_process_finding_save[713d4294-4553-4f04-bf9d-486f1b81eb0a] succeeded in 0.09412636328488588s: None
defectdojo-celery-worker-675b4c8b8b-229bd celery [20/Oct/2023 15:02:03] INFO [celery.worker.strategy:161] Task dojo.finding.helper.post_process_finding_save[d3c85038-91e8-4587-8622-0de53572ae9b] received
defectdojo-celery-worker-675b4c8b8b-229bd celery [20/Oct/2023 15:02:03] INFO [celery.app.trace:131] Task dojo.finding.helper.post_process_finding_save[d3c85038-91e8-4587-8622-0de53572ae9b] succeeded in 0.09677029121667147s: None
defectdojo-celery-worker-675b4c8b8b-229bd celery [20/Oct/2023 15:02:03] INFO [celery.worker.strategy:161] Task dojo.finding.helper.post_process_finding_save[b4039141-cf7d-42f2-86e8-e14039f91128] received
defectdojo-celery-worker-675b4c8b8b-229bd celery [20/Oct/2023 15:02:03] INFO [celery.app.trace:131] Task dojo.finding.helper.post_process_finding_save[b4039141-cf7d-42f2-86e8-e14039f91128] succeeded in 0.22523714136332273s: None
defectdojo-celery-worker-675b4c8b8b-229bd celery [20/Oct/2023 15:02:03] INFO [celery.worker.strategy:161] Task dojo.finding.helper.post_process_finding_save[8c98f6d6-e0c6-49e3-ab68-c366c2dee6ea] received
defectdojo-celery-worker-675b4c8b8b-229bd celery [20/Oct/2023 15:02:03] INFO [celery.app.trace:131] Task dojo.finding.helper.post_process_finding_save[8c98f6d6-e0c6-49e3-ab68-c366c2dee6ea] succeeded in 0.2080657733604312s: None
```

```
2023-10-20 15:02:00 INFO  DefectDojoPersistenceProvider:25 - Starting DefectDojo persistence provider
2023-10-20 15:02:01 INFO  DefectDojoPersistenceProvider:36 - Downloading Scan Result
2023-10-20 15:02:01 INFO  DefectDojoPersistenceProvider:40 - Uploading Findings to DefectDojo at: https://REDACTED
2023-10-20 15:02:01 INFO  VersionedEngagementsStrategy:85 - Running with DefectDojo User Id: 12
2023-10-20 15:02:01 INFO  VersionedEngagementsStrategy:203 - Looking for ID of ProductType 'Customer Facing Infrastructure'
2023-10-20 15:02:01 INFO  VersionedEngagementsStrategy:210 - Using ProductType Id: 7
2023-10-20 15:02:03 INFO  VersionedEngagementsStrategy:115 - Uploaded Scan Report as testID 2676 to DefectDojo
2023-10-20 15:02:03 INFO  DefectDojoPersistenceProvider:44 - Identified total Number of findings in DefectDojo: 3
2023-10-20 15:02:03 INFO  DefectDojoPersistenceProvider:51 - Overwriting secureCodeBox findings with the findings from DefectDojo.
2023-10-20 15:02:04 INFO  S3Service:25 - Uploading Findings to S3
2023-10-20 15:02:04 INFO  KubernetesService:88 - Updating Scan metadata
2023-10-20 15:02:04 INFO  DefectDojoPersistenceProvider:63 - DefectDojo Persistence Completed
```

**Comment by J12934** [comment]
Hi @moxli 
no unfortunatly the only answer I got seems to be that it is not possible to check reliably that the process is finished.

This means we have two options:

1. add this feature to DefectDojo, so that you get a id of this async process back and can you the api to wait for it to be completed
2. add a (configurable) wait between upload and refetch. Not ideal but a lot easier than option 1.

We'd be happy to accept pull requests for either solution :)

**Comment by moxli** [comment]
@J12934 I added a configurable wait to the DefectDojo Java client and tested it :) 
It seems to fix my issue 🎉
I will open up a PR in the client and this project for the changes.
I will need some feedback from your side on the best approach / spot for the wait in the hooks code.

**Comment by moxli** [comment]
Fixed in [4.4.0](https://github.com/secureCodeBox/secureCodeBox/releases/tag/v4.4.0) by https://github.com/secureCodeBox/secureCodeBox/pull/2208
